### ISSUE TITLE
mlflow.pyfunc.load_model is loading model of class <class 'mlflow.pyfunc.PyFuncModel'> instead of original class 

### ISSUE THREAD

**Issue by RATED-R-SUNDRAM** [mlflow/mlflow#12119]
Please find attached the below code where I have written a simple NN using pytorch wrapped in a class which inherits the mlflow.pyfunc.PythonModel and when i save it and laod it back getting errors at using functions of the original class (commented in the last code block), while checking the the type of model i was getting <class 'mlflow.pyfunc.PyFuncModel'> instead of original class, any suggestions why I am facing the issue and whats the fix
`

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os
import mlflow
import torch
import torch.nn as nn
from sklearn.model_selection import train_test_split 

#mlflow.create_experiment("custom_clf_pytorch")
mlflow.set_experiment("custom_clf_pytorch")
mlflow.end_run()
mlflow.start_run(run_name="custom_clf_pytorch3")

mlflow.autolog()
df= pd.read_csv(r"D:\ONE_DRIVE\Desktop\PYTORCH\iris.csv")
df['species']=df['species'].astype('category')
df['species_cat']=df['species'].cat.codes
#print(df.head)
print(df.columns)
df=df.drop(['species'],axis=1)
print(df.columns)
x= np.array(df.iloc[:,:-1])
y= np.array(df.iloc[:,-1])
y=np.eye(3)[y]
X_train, X_test, y_train, y_test = train_test_split(x,y , 
                                   random_state=104,  
                                   test_size=0.25,  
                                   shuffle=True) 
xtr= torch.tensor(X_train,dtype=torch.float)
ytr=  torch.tensor(y_train,dtype=torch.float)
xte= torch.tensor(X_test,dtype=torch.float)
yte=  torch.tensor(y_test,dtype=torch.float)
print(x,y)

print(xtr.shape,ytr.shape,xte.shape,yte.shape)

print(ytr)

class Model(nn.Module):
    def __init__(self):
        super(Model, self).__init__()
        self.fc1 = nn.Linear(4, 6)  
        self.relu = nn.ReLU()
        self.fc2 = nn.Linear(6, 3)
        self.softmax = nn.Softmax(dim=1)

    def forward(self, x):
        x = self.fc1(x)
        x = self.relu(x)
        x = self.fc2(x)
        x = self.softmax(x)
        return x

model = Model()
criterion = nn.BCELoss()
optim=torch.optim.RMSprop(model.parameters(),lr=0.05)

class custom(mlflow.pyfunc.PythonModel):
    def __init__(self,model,loss,optim):
        super(custom,self).__init__()
        self.model=model
        self.loss=loss
        self.optim=optim
        self.device=torch.device("cuda" if torch.cuda.is_available() else "cpu")
    
    def fit(self,X_train,y_train):
        epoch = 0
        while(epoch<1000):
            y_pred = self.model(X_train)
            loss = self.loss(y_pred,y_train)
            self.optim.zero_grad()
            loss.backward()
            self.optim.step()
            epoch+=1
            if(epoch%100==0):
                print("Epoch:",epoch,"Loss:",loss.item()) 
    def predict(self,X):
        return self.model(X)  # Adjusted to accept only X
    def configure_optimizers(self):
        return self.optim
    def get_model(self):
        return self.model
    
clf = custom(model,criterion,optim)
clf.fit(xtr,ytr)

mlflow.pyfunc.log_model(artifact_path="model",python_model=clf)

clf2 = mlflow.pyfunc.load_model("runs:/a8fc04526db8406f8778e06487a749b1/model")

clf2.predict(xte)  # getting error that custom.predicts takes 2 argument but 3 was given
print(type(clf2)) # getting 'mlflow.pyfunc.PyFuncModel'> instead of original class

`

**Comment by daniellok-db** [comment]
If you want to access the original logged model, you can call the `unwrap_python_model()` function ([link to docs](https://mlflow.org/docs/latest/python_api/mlflow.pyfunc.html#mlflow.pyfunc.PyFuncModel.unwrap_python_model)).

For example:
```
loaded_model = mlflow.pyfunc.load_model("runs:/a8fc04526db8406f8778e06487a749b1/model")
clf2 = loaded_model.unwrap_python_model()
```

It also looks like your pyfunc model doesn't properly implement the interface required for classes that inherit from `PythonModel` (though it's a bit hard to tell due to the formatting of the issue summary). For example, `predict()` should have the following signature:

```
def predict(self, context, model_input, params=None):
  ...
```

For more info, check out the tutorial on building custom pyfunc models ([link to docs](https://mlflow.org/docs/latest/traditional-ml/creating-custom-pyfunc/notebooks/introduction.html)).

**Comment by B-Step62** [comment]
Closing as this is an intended behavior.
